core/dbt/include/global_project/macros/materializations/common/merge.sql

```diff
@@ -19,19 +19,35 @@
     {{ return(dest_cols_csv) }}
 {% endmacro %}
 
-
-{% macro common_get_merge_sql(target, source, unique_key, dest_columns) -%}
+{% macro common_get_merge_sql(target, source, unique_key, dest_columns, partition_by) -%}
     {%- set dest_cols_csv =  get_quoted_csv(dest_columns | map(attribute="name")) -%}
 
-    merge into {{ target }} as DBT_INTERNAL_DEST
-    using {{ source }} as DBT_INTERNAL_SOURCE
+    {%- set conditions = [] -%}
 
     {% if unique_key %}
-        on DBT_INTERNAL_SOURCE.{{ unique_key }} = DBT_INTERNAL_DEST.{{ unique_key }}
-    {% else %}
-        on FALSE
+        {% set unique_key_match %}
+            DBT_INTERNAL_SOURCE.{{ unique_key }} = DBT_INTERNAL_DEST.{{ unique_key }}
+        {% endset %}
+        {% do conditions.append(unique_key_match) %}
+    {% endif %}
+    
+    {# BigQuery only #}
+    {%- if partition_by -%}
```

Comment:
I don't love that we're encoding this BigQuery specific logic in the common merge implementation. Further, the `pprint_partition_field` macro used below is only defined in the BigQuery plugin. This probably won't be an issue, as the Snowflake code path won't be providing a `partition_by` arg here, but it does feel out of place to me.

I'd be comfortable just leaving the existing `common_get_merge_sql` as-is and instead baking this logic directly into the BigQuery implementation of `get_merge_sql`:

https://github.com/fishtown-analytics/dbt/blob/e080bfc79ab11a4d145de4c8da9d955c5e136d92/plugins/bigquery/dbt/include/bigquery/macros/materializations/merge.sql#L1-L3

Alternatively, we could make this a little bit more generic and change the `partition_by` arg for this macro to `predicates` (or similar). The BigQuery implementation of `get_merge_sql` could supply a list of predicates to this macro, but other implementations (like Snowflake) could provide an empty list.

Either way, let's move the partitioning-specific logic out of this macro.